/*
 * Copyright (c) 2018, 2019 by Delphix. All rights reserved.
 */

syntax = "proto3";

option java_multiple_files = true;

package com.delphix.virtualization.common;

message RemoteConnection {
  RemoteEnvironment environment = 1;
  RemoteUser user = 2;
}

message RemoteEnvironment {
  string name = 1;
  string reference = 2;
  RemoteHost host = 3;
}

message RemoteHost {
  string name = 1;
  string reference = 2;
  string binary_path = 3;
  string scratch_path = 4;
}

message RemoteUser {
  string name = 1;
  string reference = 2;
}

message LinkedSource {
  string guid = 1;
  PluginDefinedObject parameters = 2;
}

message DirectSource {
  RemoteConnection connection = 1;
  LinkedSource linked_source = 2;
}

message StagedSource {
  LinkedSource linked_source = 1;
  RemoteConnection connection = 3;
  SingleMount staged_mount = 4;
}

message SingleMount {
  RemoteEnvironment remote_environment = 1;
  string mount_path = 2;
  string shared_path = 3; // not supported on Windows
}

message VirtualSource {
  string guid = 1;
  RemoteConnection connection = 2;
  SingleMount primary_mount = 3;
  repeated SingleMount additional_mounts = 4;
  PluginDefinedObject parameters = 5;
}

message SourceConfig {
  string name = 1;
  PluginDefinedObject parameters = 2;
}

message Repository {
  string name = 1;
  PluginDefinedObject parameters = 2;
}

message Snapshot {
  string name = 1;
  PluginDefinedObject parameters = 2;
}

message PluginDefinedObject {
  string json = 1;
}

message OwnershipSpec {
  int32 uid = 1;
  int32 gid = 2;
}
